### ISSUE TITLE
filling array using `x = Array(y).map(Number.prototype.valueOf,0)` fails with TypeError: Cannot find default value for object.

### ISSUE THREAD

**Issue by tonescones** [mozilla/rhino#433]
an expression like `x = Array(y).map(Number.prototype.valueOf,0);` started to fail after release 1.7.7

ActionError: wrapped org.mozilla.javascript.EcmaError: TypeError: Cannot find default value for object.

**Comment by gbrail** [comment]
It seems to work for me. Do you have a more complete example that we can run so that we can see the failure happening? Thanks!

**Comment by tonescones** [comment]
js> var foo = Array.apply(null, Array(5)).map(Number.prototype.valueOf, 0);
js: uncaught JavaScript runtime exception: TypeError: Cannot find default value for object.

On the other hand, if you replace 'null' with foo, it works fine.
js> var foo = Array.apply(null, Array(5)).map(Number.prototype.valueOf, 0);
0,0,0,0,0

**Comment by tonescones** [comment]
@gbrail btw is there a way for me to contact you directly?

**Comment by gbrail** [comment]
Yes, I see that too on the latest code base. It seems to be something with a null argument to Array.apply.

A first step would be for someone to take a look at the Array class and see what changed -- I know that we made a bunch of changes around null handling a while back...

java -jar ~/src/rhino/buildGradle/libs/rhino-1.7.11-SNAPSHOT.jar 
Rhino 1.7.11-SNAPSHOT 2018 05 01
js> var a = Array(5);
js> a[0] = 'hello';
hello
js> a[1] = 'goodbye';
goodbye
js> Array.apply(undefined, a).map(function(a) { print (a); });
hello
goodbye
undefined
undefined
undefined
,,,,
js> Array.apply(null, a).map(function(a) { print (a); });
js: uncaught JavaScript runtime exception: TypeError: Cannot find default value for object.

**Comment by gbrail** [comment]
Well, this is a bug from 2018 but that doesn't mean we shouldn't fix it!

I think that the root cause is that in the constructor code in the NativeArray class, it uses "null" for "this" to determine whether it's being called directly or via "new", even though that's clearly not the case.

I think that the best way to solve this and many other problems is to make NativeArray use lambda functions rather than the complexity of IdScriptableObject. I don't know if that's too risky for a release that we should ship very soon but it may be worth a try.

**Comment by gbrail** [comment]
I'm quite sure that we can fix this (and a bunch of other things too) by switching NativeArray from an IdScriptableObject to a LambdaConstructor-based implementation. We have started to do that for other native classes too.

However, I spent a few hours on this and it's not as easy as it is for some other classes due to the unique nature of arrays in JavaScript. (Getting tired of saying that, since "the unique nature of foos in JavaScript" just applies to every single thing in the language it seems.)

In other words, I think that we can fix this but I would like to get a release out soon and I don't have time for both things until next year, so I'd rather do the release and take the time to fix arrays carefully.
